cql3/statements/select_statement.cc

```diff
@@ -283,31 +283,48 @@ select_statement::make_partition_slice(const query_options& options) const
         std::reverse(bounds.begin(), bounds.end());
         ++_stats.reverse_queries;
     }
+
+    const uint64_t per_partition_limit = get_per_partition_limit_with_default_and_unset_guard(
+        get_limit(options, _per_partition_limit), options);
     return query::partition_slice(std::move(bounds),
-        std::move(static_columns), std::move(regular_columns), _opts, nullptr, get_per_partition_limit(options));
+        std::move(static_columns), std::move(regular_columns), _opts, nullptr, per_partition_limit);
 }
 
-uint64_t select_statement::do_get_limit(const query_options& options,
-                                        const std::optional<expr::expression>& limit,
-                                        const expr::unset_bind_variable_guard& limit_unset_guard,
-                                        uint64_t default_limit) const {
-    if (!limit.has_value() || limit_unset_guard.is_unset(options) || _selection->is_aggregate()) {
-        return default_limit;
-    }
-
-    auto val = expr::evaluate(*limit, options);
-    if (val.is_null()) {
-        throw exceptions::invalid_request_exception("Invalid null value of limit");
+select_statement::get_limit_result select_statement::get_limit(
+    const query_options& options, const std::optional<expr::expression>& limit) const
+{
+    if (!limit.has_value()) {
+        return bo::success(query::max_rows);
     }
     try {
+        auto val = expr::evaluate(*limit, options);
+        if (val.is_null()) {
+            return bo::failure(exceptions::invalid_request_exception("Invalid null value of limit"));
+        }
         auto l = val.view().validate_and_deserialize<int32_t>(*int32_type);
         if (l <= 0) {
-            throw exceptions::invalid_request_exception("LIMIT must be strictly positive");
+            return bo::failure(exceptions::invalid_request_exception("LIMIT must be strictly positive"));
         }
-        return l;
+        return bo::success(l);
     } catch (const marshal_exception& e) {
-        throw exceptions::invalid_request_exception("Invalid limit value");
+        return bo::failure(exceptions::invalid_request_exception("Invalid limit value"));
+    } catch (const exceptions::invalid_request_exception& e) {
+        return bo::failure(e);
+    }
+}
+
+uint64_t select_statement::get_limit_with_default_and_unset_guard(const get_limit_result& limit, const query_options& options) const {
+    if (!limit.has_value() || _limit_unset_guard.is_unset(options) || _selection->is_aggregate()) {
+        return query::max_rows;
+    }
+    return limit.value();
+}
+
+uint64_t select_statement::get_per_partition_limit_with_default_and_unset_guard(const get_limit_result& limit, const query_options& options) const {
```

Comment:
This function confused me the most. It doesn't "get per partition limit", actually it usually the limit passed to it as a parameter (which is LIMIT not PER PARTITION LIMIT, right?) and only if unset, it returns query::partition_max_rows, which unless I'm missing something, is just a constant maximum uint64_t, nothing to do with partition limit.